src/tribler/core/components/knowledge/db/knowledge_db.py

```diff
@@ -206,134 +206,173 @@ def add_auto_generated(self, subject_type: ResourceType, subject: str, predicate
         """ Get all suggestions for a particular subject.
 
         Args:
+            subject_type: a type of the subject.
             subject: a string that represents the subject.
             predicate: the enum that represents a predicate of querying operations.
-            case_sensitive: if True, then Resources will be selected in case sensitive manner. if False, then Resources
-                will be selected in case insensitive manner.
+            case_sensitive: if True, then Resources are selected in a case-sensitive manner. if False, then Resources
+                are selected in a case-insensitive manner.
 
         Returns: a list of the strings representing the objects.
         """
         self.logger.debug(f"Getting suggestions for {subject} with {predicate}")
 
-        def show_suggestions_condition(statement):
-            return not statement.local_operation and \
-                   between(statement.score, HIDE_THRESHOLD + 1, SHOW_THRESHOLD - 1)
-
-        return self._get_resources(subject, show_suggestions_condition, predicate, case_sensitive,
-                                   is_normal_direction=True)
+        suggestions = self.get_objects(
+            subject_type=subject_type,
+            subject=subject,
+            predicate=predicate,
+            case_sensitive=case_sensitive,
+            condition=lambda s: not s.local_operation and between(s.score, HIDE_THRESHOLD + 1, SHOW_THRESHOLD - 1)
+        )
+        return suggestions
 
-    def get_subjects_intersection(self, objects: Set[str], predicate: ResourceType,
+    def get_subjects_intersection(self, subjects_type: Optional[ResourceType], objects: Set[str],
+                                  predicate: Optional[ResourceType],
                                   case_sensitive: bool = True) -> Set[str]:
         """Queries the subjects with the given objects and the predicate. Then made an intersection among them.
 
         In the Tribler, this method is mostly used for searching by tags.
 
         Args:
+            subjects_type: a type of subjects.
             objects: a set of strings that represents the objects.
             predicate: the enum that represents a predicate of querying operations.
-            case_sensitive: if True, then Resources will be selected in case sensitive manner. if False, then Resources
-                will be selected in case insensitive manner.
+            case_sensitive: if True, then Resources are selected in a case-sensitive manner. if False, then Resources
+                are selected in a case-insensitive manner.
 
         Returns: a list of the strings representing the subjects.
         """
-        # FIXME: Ask @kozlovsky how to do it in a proper way
-        sets = [set(self.get_subjects(o, predicate, case_sensitive)) for o in objects]
+        sets = [set(self.get_subjects(subjects_type, predicate, o, case_sensitive)) for o in objects]
```

Comment:
The PR implementation of `get_subject_intersection` will work inefficiently for a real-size database because it performs some unnecessary steps:
1. Join statements with subject resources to obtain subject names
2. Fully construct subject resources as Python objects
3. Perform intersection of subject names instead of subject ids

The following implementation should be much faster. Unfortunately, it was necessary to inline some expressions to make it work:
```python
    def get_subjects_with_objects(self, object_names: Set[str], subject_type: ResourceType, object_type: ResourceType,
                                  case_sensitive: bool = True) -> Set[str]:
        if case_sensitive:
            def name_condition(obj, obj_name):
                return obj.name == obj_name
        else:
            def name_condition(obj, obj_name):
                return obj.name.lower() == obj_name.lower()

        query = select(subject.name for subject in self.instance.Resource if subject.type == subject_type.value)
        for object_name in object_names:
            query = query.where(lambda subject: subject in (
                s.subject for s in self.instance.Statement
                if s.local_operation == Operation.ADD.value or not s.local_operation and s.score >= SHOW_THRESHOLD
                and s.object in (
                    object for object in self.instance.Resource
                    if object.type == object_type.value
                    and name_condition(object, object_name)
                )
            ))
        return set(query)

    def get_torrents_with_tags(self, tag_names: Set[str]):
        return self.get_subjects_with_objects(tag_names, ResourceType.TORRENT, ResourceType.TAG)
```

It is more efficient for the following reasons:
1. It performs the intersection directly in SQL without constructing Python objects;
2. It intersects subject ids instead of subject names, which is more efficient as joining with the Resource table to obtain subject names is performed only after the intersection and for a potentially much smaller number of objects;
3. Only resource names are loaded into Python without constructing full Resource objects in memory.

Reply:
Thank you! Added!

Reply:
I've implemented this suggestion.
But the following error started to occur:

https://github.com/Tribler/tribler/actions/runs/3460338705/jobs/5776779808#step:7:273

```python
src/tribler/core/components/metadata_store/remote_query_community/tests/test_remote_search_by_tags.py . [ 54%]
.............................F

=================================== FAILURES ===================================
_____________________ TestTagDB.test_show_local_resources ______________________

self = <tribler.core.components.knowledge.db.tests.test_knowledge_db.TestTagDB testMethod=test_show_local_resources>

    @db_session
    def test_show_local_resources(self):
        # Test that locally added tags have a priority to show.
        # That means no matter of other peers opinions, locally added tag should be visible.
        self.add_operation(self.db, ResourceType.TORRENT, 'infohash1', ResourceType.TAG, 'tag1', b'peer1',
                           operation=Operation.REMOVE)
        self.add_operation(self.db, ResourceType.TORRENT, 'infohash1', ResourceType.TAG, 'tag1', b'peer2',
                           operation=Operation.REMOVE)
>       assert not self.db.get_objects(subject='infohash1', predicate=ResourceType.TAG)

src/tribler/core/components/knowledge/db/tests/test_knowledge_db.py:279: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/tribler/core/components/knowledge/db/knowledge_db.py:280: in get_objects
    return [s.object.name for s in statements]
src/tribler/core/components/knowledge/db/knowledge_db.py:280: in <listcomp>
    return [s.object.name for s in statements]
src/tribler/core/components/knowledge/db/knowledge_db.py:243: in _get_statements
    results = orm.select(_ for _ in statements_getter(resource)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/core.py:3583: in select
    query = query.filter(func, globals, locals)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/core.py:6114: in filter
    return query._process_lambda(func, globals, locals, order_by=False)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/core.py:6083: in _process_lambda
    new_translator = prev_translator.apply_lambda(func_id, new_filter_num, order_by, func_ast, argnames, original_names, extractors, new_vars, vartypes)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:922: in apply_lambda
    translator.dispatch(func_ast)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:107: in dispatch
    return ASTTranslator.dispatch(translator, node)  # default route
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/asttranslation.py:40: in dispatch
    translator.dispatch(child)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:107: in dispatch
    return ASTTranslator.dispatch(translator, node)  # default route
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/asttranslation.py:40: in dispatch
    translator.dispatch(child)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:107: in dispatch
    return ASTTranslator.dispatch(translator, node)  # default route
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/asttranslation.py:36: in dispatch
    stop = translator.call(pre_method, node)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:195: in call
    reraise(exc_class, exc, tb)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/utils/utils.py:88: in reraise
    try: raise exc.with_traceback(tb)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:184: in call
    monad = method(translator, node)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:967: in preCompare
    translator.dispatch(left)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:107: in dispatch
    return ASTTranslator.dispatch(translator, node)  # default route
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/asttranslation.py:47: in dispatch
    translator.call(post_method, node)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:195: in call
    reraise(exc_class, exc, tb)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/utils/utils.py:88: in reraise
    try: raise exc.with_traceback(tb)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:184: in call
    monad = method(translator, node)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:1044: in postAttribute
    return node.value.monad.getattr(node.attr)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:1450: in wrapper
    return method(monad, *args, **kwargs)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:2175: in getattr
    return new_monad()
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:1450: in wrapper
    return method(monad, *args, **kwargs)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:[268](https://github.com/Tribler/tribler/actions/runs/3460338705/jobs/5776779808#step:7:271)8: in __call__
    translator.dispatch(func_ast)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:107: in dispatch
    return ASTTranslator.dispatch(translator, node)  # default route
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/asttranslation.py:36: in dispatch
    stop = translator.call(pre_method, node)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:195: in call
    reraise(exc_class, exc, tb)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/utils/utils.py:88: in reraise
    try: raise exc.with_traceback(tb)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:184: in call
    monad = method(translator, node)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:967: in preCompare
    translator.dispatch(left)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:107: in dispatch
    return ASTTranslator.dispatch(translator, node)  # default route
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/asttranslation.py:40: in dispatch
    translator.dispatch(child)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:107: in dispatch
    return ASTTranslator.dispatch(translator, node)  # default route
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/asttranslation.py:47: in dispatch
    translator.call(post_method, node)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:195: in call
    reraise(exc_class, exc, tb)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/utils/utils.py:88: in reraise
    try: raise exc.with_traceback(tb)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:184: in call
    monad = method(translator, node)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:1010: in postName
    monad = translator.resolve_name(node.id)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:1015: in resolve_name
    throw(TranslationError, 'Name %s is not found in %s' % (name, translator.namespace))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

exc_type = <class 'pony.orm.asttranslation.TranslationError'>
args = ("Name obj is not found in {'self': <pony.orm.sqltranslation.ObjectIterMonad object at 0x7f78c223a7f0>}",)
kwargs = {}

    def throw(exc_type, *args, **kwargs):
        if isinstance(exc_type, Exception):
            assert not args and not kwargs
            exc = exc_type
        else: exc = exc_type(*args, **kwargs)
        exc.__cause__ = None
        try:
            if not (pony.MODE == 'INTERACTIVE' and options.CUT_TRACEBACK):
>               raise exc
E               pony.orm.asttranslation.TranslationError: Name obj is not found in {'self': <pony.orm.sqltranslation.ObjectIterMonad object at 0x7f78c223a7f0>} (inside Statement.score)

../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/utils/utils.py:99: TranslationErrorsrc/tribler/core/components/metadata_store/remote_query_community/tests/test_remote_search_by_tags.py . [ 54%]
.............................F

=================================== FAILURES ===================================
_____________________ TestTagDB.test_show_local_resources ______________________

self = <tribler.core.components.knowledge.db.tests.test_knowledge_db.TestTagDB testMethod=test_show_local_resources>

    @db_session
    def test_show_local_resources(self):
        # Test that locally added tags have a priority to show.
        # That means no matter of other peers opinions, locally added tag should be visible.
        self.add_operation(self.db, ResourceType.TORRENT, 'infohash1', ResourceType.TAG, 'tag1', b'peer1',
                           operation=Operation.REMOVE)
        self.add_operation(self.db, ResourceType.TORRENT, 'infohash1', ResourceType.TAG, 'tag1', b'peer2',
                           operation=Operation.REMOVE)
>       assert not self.db.get_objects(subject='infohash1', predicate=ResourceType.TAG)

src/tribler/core/components/knowledge/db/tests/test_knowledge_db.py:279: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/tribler/core/components/knowledge/db/knowledge_db.py:280: in get_objects
    return [s.object.name for s in statements]
src/tribler/core/components/knowledge/db/knowledge_db.py:280: in <listcomp>
    return [s.object.name for s in statements]
src/tribler/core/components/knowledge/db/knowledge_db.py:243: in _get_statements
    results = orm.select(_ for _ in statements_getter(resource)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/core.py:3583: in select
    query = query.filter(func, globals, locals)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/core.py:6114: in filter
    return query._process_lambda(func, globals, locals, order_by=False)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/core.py:6083: in _process_lambda
    new_translator = prev_translator.apply_lambda(func_id, new_filter_num, order_by, func_ast, argnames, original_names, extractors, new_vars, vartypes)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:922: in apply_lambda
    translator.dispatch(func_ast)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:107: in dispatch
    return ASTTranslator.dispatch(translator, node)  # default route
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/asttranslation.py:40: in dispatch
    translator.dispatch(child)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:107: in dispatch
    return ASTTranslator.dispatch(translator, node)  # default route
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/asttranslation.py:40: in dispatch
    translator.dispatch(child)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:107: in dispatch
    return ASTTranslator.dispatch(translator, node)  # default route
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/asttranslation.py:36: in dispatch
    stop = translator.call(pre_method, node)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:195: in call
    reraise(exc_class, exc, tb)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/utils/utils.py:88: in reraise
    try: raise exc.with_traceback(tb)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:184: in call
    monad = method(translator, node)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:967: in preCompare
    translator.dispatch(left)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:107: in dispatch
    return ASTTranslator.dispatch(translator, node)  # default route
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/asttranslation.py:47: in dispatch
    translator.call(post_method, node)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:195: in call
    reraise(exc_class, exc, tb)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/utils/utils.py:88: in reraise
    try: raise exc.with_traceback(tb)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:184: in call
    monad = method(translator, node)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:1044: in postAttribute
    return node.value.monad.getattr(node.attr)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:1450: in wrapper
    return method(monad, *args, **kwargs)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:2175: in getattr
    return new_monad()
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:1450: in wrapper
    return method(monad, *args, **kwargs)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:[268](https://github.com/Tribler/tribler/actions/runs/3460338705/jobs/5776779808#step:7:271)8: in __call__
    translator.dispatch(func_ast)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:107: in dispatch
    return ASTTranslator.dispatch(translator, node)  # default route
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/asttranslation.py:36: in dispatch
    stop = translator.call(pre_method, node)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:195: in call
    reraise(exc_class, exc, tb)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/utils/utils.py:88: in reraise
    try: raise exc.with_traceback(tb)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:184: in call
    monad = method(translator, node)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:967: in preCompare
    translator.dispatch(left)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:107: in dispatch
    return ASTTranslator.dispatch(translator, node)  # default route
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/asttranslation.py:40: in dispatch
    translator.dispatch(child)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:107: in dispatch
    return ASTTranslator.dispatch(translator, node)  # default route
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/asttranslation.py:47: in dispatch
    translator.call(post_method, node)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:195: in call
    reraise(exc_class, exc, tb)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/utils/utils.py:88: in reraise
    try: raise exc.with_traceback(tb)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:184: in call
    monad = method(translator, node)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:1010: in postName
    monad = translator.resolve_name(node.id)
../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/orm/sqltranslation.py:1015: in resolve_name
    throw(TranslationError, 'Name %s is not found in %s' % (name, translator.namespace))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

exc_type = <class 'pony.orm.asttranslation.TranslationError'>
args = ("Name obj is not found in {'self': <pony.orm.sqltranslation.ObjectIterMonad object at 0x7f78c223a7f0>}",)
kwargs = {}

    def throw(exc_type, *args, **kwargs):
        if isinstance(exc_type, Exception):
            assert not args and not kwargs
            exc = exc_type
        else: exc = exc_type(*args, **kwargs)
        exc.__cause__ = None
        try:
            if not (pony.MODE == 'INTERACTIVE' and options.CUT_TRACEBACK):
>               raise exc
E               pony.orm.asttranslation.TranslationError: Name obj is not found in {'self': <pony.orm.sqltranslation.ObjectIterMonad object at 0x7f78c223a7f0>} (inside Statement.score)

../../../.virtualenvs/.venv/lib/python3.8/site-packages/pony/utils/utils.py:99: TranslationError
```

Reply:
Could be related to the issue printed above:
https://app.codacy.com/gh/Tribler/tribler/file/78887108477/issues/source?bid=28604826&fileBranchId=32051691#l391